Add performance unit test for Flow Exporter

zyiou opened the pull request:
This commit adds performance benchmarking for Flow Exporter. It evaluates Export() function under different number of conntrack connections, dying connections, idle records, deny connections and idle deny connections. A local server will receive the records and count number. CPU and memory profile is collected and visualized using pprof.

Also from benchmarking, we discovered and removed redundant calls like `GetNodeName()`, which is called every time when exporting a record and `ResetConnStatsWithoutLock`, which unnecessarily calls `NewConnectionKey` each time.

With fixes and changes in go-ipfix [v0.5.3](https://github.com/vmware/go-ipfix/releases/tag/v0.5.3) and redundant calls removal in this PR, we had following improvements:
Before:
```
BenchmarkExportConntrackConns   	  50	  21747542 ns/op	 3708730 B/op	   80838 allocs/op
BenchmarkExportDenyConns      	          126	  14497807 ns/op	 2522699 B/op	   53345 allocs/op
BenchmarkPoll   	                 141	   8372127 ns/op	  813834 B/op	   52738 allocs/op
```
After
```
BenchmarkExportConntrackConns   	 126	  11444126 ns/op	  459816 B/op	   18332 allocs/op
BenchmarkExportDenyConns   	     223	   6391654 ns/op	  261669 B/op	   10637 allocs/op
BenchmarkPoll   	         130	   8280194 ns/op	  818076 B/op	   52814 allocs/op
```
Improvement
| Test   |  ns/op | B/op   | allocs/op |
| ------------------------- | --------- | --- | --- |
| BenchmarkExportConntrackConns  |  reduce by 47.4%	|  reduce by 87.6% | reduce by 77.3% | 
| BenchmarkExportDenyConns  |   reduce by 42.9%	|  reduce by 89.6%	|   reduce by 80.0% | 
| BenchmarkPoll  |  less than 1% diff	|  less than 1% diff	| less than 1% diff  |

zyiou commented:
Sample metrics:
```
TOTAL_CONNECTIONS:   123551	
INIT_CONNECTIONS:   100000
UPDATE_CONNECTIONS/poll:   80000
NEW_CONNECTIONS/poll:   10000
DYING_CONNECTIONS/poll:   5000
POLL_INTERVAL(s):  1
TEST_DURATION(s):  10
MEMORY(M):  428 
```

Benchmark result:
```
go test -test.v -run=BenchmarkExport -test.benchmem -bench=. -memprofile memprofile.out -cpuprofile profile.out

goos: linux
goarch: amd64
pkg: antrea.io/antrea/pkg/agent/flowexporter/exporter
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
BenchmarkExport
BenchmarkExport-2   	       1	35014186189 ns/op	1322987648 B/op	27009979 allocs/op
```
```
go tool pprof memprofile.out
File: exporter.test
Type: alloc_space
Time: May 18, 2021 at 11:43pm (UTC)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 1103.48MB, 86.50% of 1275.64MB total
Dropped 63 nodes (cum <= 6.38MB)
Showing top 10 nodes out of 50
      flat  flat%   sum%        cum   cum%
     480MB 37.63% 37.63%   783.52MB 61.42%  github.com/vmware/go-ipfix/pkg/entities.(*dataRecord).AddInfoElement
  116.50MB  9.13% 46.76%   210.52MB 16.50%  github.com/vmware/go-ipfix/pkg/entities.EncodeToIEDataType
  105.63MB  8.28% 55.04%   108.13MB  8.48%  antrea.io/antrea/pkg/agent/flowexporter/flowrecords.(*FlowRecords).AddOrUpdateFlowRecord
      93MB  7.29% 62.33%       93MB  7.29%  github.com/vmware/go-ipfix/pkg/entities.NewInfoElementWithValue (inline)
   66.01MB  5.17% 67.51%    66.01MB  5.17%  k8s.io/klog/v2.(*loggingT).header
   64.88MB  5.09% 72.59%   140.39MB 11.01%  antrea.io/antrea/pkg/agent/flowexporter/connections.(*connectionStore).addOrUpdateConn
   61.02MB  4.78% 77.38%    61.02MB  4.78%  bytes.makeSlice
   44.42MB  3.48% 80.86%    44.42MB  3.48%  antrea.io/antrea/pkg/agent/flowexporter/exporter.getUpdateConnections
   40.52MB  3.18% 84.03%    47.02MB  3.69%  antrea.io/antrea/pkg/agent/flowexporter/exporter.getConnections
   31.50MB  2.47% 86.50%    31.50MB  2.47%  net.IP.String
```

zyiou commented:
We are having several performance improvements on go-ipfix side. Here is the performance difference for this [change](https://github.com/vmware/go-ipfix/pull/204)

Before change:
```
go test -test.v -run=BenchmarkExport -test.benchmem -bench=. -memprofile memprofile.out -cpuprofile profile.out

goos: linux
goarch: amd64
pkg: antrea.io/antrea/pkg/agent/flowexporter/exporter
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
BenchmarkExport
BenchmarkExport-2   	       1	20001730232 ns/op	1107956184 B/op	21170436 allocs/op
PASS
ok  	antrea.io/antrea/pkg/agent/flowexporter/exporter	20.131s
```
After change:
```
go test -test.v -run=BenchmarkExport -test.benchmem -bench=. -memprofile memprofile.out -cpuprofile profile.out
goos: linux
goarch: amd64
pkg: antrea.io/antrea/pkg/agent/flowexporter/exporter
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
BenchmarkExport
BenchmarkExport-2   	       1	20002124526 ns/op	801765208 B/op	13902598 allocs/op
PASS
ok  	antrea.io/antrea/pkg/agent/flowexporter/exporter	20.132s
```

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,269 @@
+	}
+	// Generate connections for Dumpflows, each round with update connections, new connections and dying connections
+	mockConnDumper.EXPECT().DumpFlows(testZoneFilter).Return(conns, testConnectionsCount, nil).Times(1)
+	for i := 0; i < 5; i++ {
```
what does 5 mean here? Are we polling only five times during the test?

zyiou replied:
Yes in the previous setting. Changed to testduration/pollinterval to make it clear.

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,269 @@
+		b.Fatalf("Got error when creating a local server: %v", err)
+	}
+	prevCount := 0
+	ticker := time.NewTicker(testIdleTimeOut)
```
use timer here?

zyiou replied:
If we use timer instead of ticker, we will need to reset timer each time. Is there any special reason for timer?

srikartati replied:
As we are sleeping for `testDuration`, I thought this ticker is only expected to trigger once as a timeout like the name suggests `testIdleTimeOut`. Looks like to update `prevCount`, you want this to be a periodic ticker. Could we just timeout once without worrying about how many messages are received?

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,269 @@
+
+const (
+	testPollInterval           = 2 * time.Second
+	testConnectionsCount       = 30000
```
Looking at the other counts, it seems like you consider 9K connections to stay the same (30K - (20K +1K)). I feel they need to be updated (at least times) or die.

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,269 @@
+
+	var maxAlloc uint64
+	go statMaxMemAlloc(&maxAlloc, 500*time.Millisecond, stopChan2)
+	go connStore.Run(stopChan1)
+	go exp.Run(stopChan2)
```
why separate channels?

zyiou replied:
I want the poll to be finished after test duration and close exporter when exporter does not receive records for some time (`testIdleTimeOut`). That's why we need two separate channels to stop connection store and exporter, respectively.

srikartati replied:
>... close exporter when exporter does not receive records for some time (testIdleTimeOut)

Do you mean the collector could not receive the records? The exporter just sends the records in the test here.
Any particular reason to stop polling? 
In other words, do we want to wait till we receive all messages at the collector? Could we just do the duration and close the collector/exporting processes when it finishes? We just record the number of messages at the collector before closing the process. 
This way we can see within specific duration, how many messages are collected. As CPU time is optimized, we could see an increase in the number of messages at the collector. What do you think of this approach?

zyiou replied:
Got it. Yes this approach makes sense to me. As you mentioned, I was planning to wait for all messages to safely close the exporting process.

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,269 @@
+	testNewConnectionsCount    = 4000
+	testDyingConnectionsCount  = 1000
+	testZoneFilter             = uint16(65520)
+	testDuration               = 15 * time.Second
```
It's better to have the duration in minutes to stress the flow exporter.

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,269 @@
+	testDyingConnectionsCount  = 1000
+	testZoneFilter             = uint16(65520)
+	testDuration               = 15 * time.Second
+	testIdleTimeOut            = 5 * time.Second
```
is this timeout for the exporter channel?

zyiou replied:
Yes

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,269 @@
+)
+
+const (
+	testPollInterval           = 2 * time.Second
```
Let's have poll interval less than active export timeout

srikartati reviewed (commented):
In addition to the metrics, can we make a check if our memory utilization is linear with number of new connections added. Basically, the idea is to detect the memory leaks.

zyiou commented:
> In addition to the metrics, can we make a check if our memory utilization is linear with number of new connections added. Basically, the idea is to detect the memory leaks.

Sure. Will create a larger vm for testing.

srikartati commented:
> > In addition to the metrics, can we make a check if our memory utilization is linear with number of new connections added. Basically, the idea is to detect the memory leaks.
> 
> Sure. Will create a larger vm for testing.

I just meant to change the initial number of connections through an input parameter (from a lower number to a higher number) and tracking memory utilization. Current number can be the maximum number.

zyiou commented:
Test results with different number of new connections per poll:
| # new conns / poll  |   B/op   | allocs/op | max memory alloc(M) | 
| ------------------------- | ------- | --------- | ----- |
| 2000 | 1,872,791,432  | 38,548,557 | 431  | 
| 4000 | 2,525,452,400  | 52,035,182 | 576 | 
| 6000 | 3,076,454,200  | 63,339,188 | 726  | 
| 8000 | 3,730,222,696  | 77,111,198 | 899  | 
| 10000 | 4,432,394,112  | 90,846,552 | 1020  | 

Memory consumption is roughly linear as number of new connections increase.

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,245 @@
+	return connections
+}
+
+func statMaxMemAlloc(maxAlloc *uint64, interval time.Duration, stopCh chan struct{}) {
+	var memStats goruntime.MemStats
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			goruntime.ReadMemStats(&memStats)
+			if memStats.Alloc > *maxAlloc {
+				*maxAlloc = memStats.Alloc
+			}
+		case <-stopCh:
+			return
+		}
+	}
+}
```
do we still need this memstat collection? Are these something different from what we are getting from benchmark test?

zyiou replied:
No we don't need this any more. Have it removed. Thanks!

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,245 @@
+
+var count = 0
+
+func BenchmarkExport(b *testing.B) {
```
Better to add the latest result in the comment here, which can be used as a reference for future runs.

srikartati reviewed (commented):
Thanks for doing comparison by increasing the connection counts. Can that parameter be passed into the benchmark test using the go test command?

antoninbas reviewed (commented):
Sorry for the late review. Overall I am a bit confused by the format of this benchmark. Because we do not include the actual conntrack polling in the benchmark at all, it's not really an "end-to-end" benchmark of the FlowExporter. Given that, I think it would have been more convenient to write distinct benchmarks (using an actual Go benchmark) for the `ConntrackConnectionStore` (`Poll()` method) and the `flowExporter` (`export()` method). Then we would have a good idea of the performance of each one and I feel like the CPU / memory data would be more accurate (as the benchmarked function would be invoked multiple times).

zyiou commented:
Updated testing result:
```
go test -test.v -run=BenchmarkExport -test.benchmem -bench=. -memprofile memprofile.out -cpuprofile profile.out
goos: linux
goarch: amd64
pkg: antrea.io/antrea/pkg/agent/flowexporter/exporter
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
BenchmarkExport
    exporter_perf_test.go:61: 
        Summary:
        Number of conntrack connections: 20000
        Number of dying conntrack connections: 10000
        Number of deny connections: 20000
        Number of idle deny connections: 10000
        Total connections received: 37788
BenchmarkExport-2   	       1	1123878204 ns/op	135576880 B/op	 3458292 allocs/op
PASS
ok  	antrea.io/antrea/pkg/agent/flowexporter/exporter	1.298s
```
```
go tool pprof memprofile.out
File: exporter.test
Type: alloc_space
Time: Jun 22, 2021 at 2:01am (UTC)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 103.32MB, 83.29% of 124.05MB total
Dropped 16 nodes (cum <= 0.62MB)
Showing top 10 nodes out of 68
      flat  flat%   sum%        cum   cum%
   22.50MB 18.14% 18.14%       27MB 21.77%  github.com/vmware/go-ipfix/pkg/entities.(*dataRecord).AddInfoElement
   15.76MB 12.70% 30.84%    15.76MB 12.70%  antrea.io/antrea/pkg/agent/flowexporter/flowrecords.(*FlowRecords).AddFlowRecordToMap
   15.50MB 12.50% 43.34%    15.50MB 12.50%  k8s.io/klog/v2.(*loggingT).header
   12.04MB  9.71% 53.05%    12.04MB  9.71%  antrea.io/antrea/pkg/agent/flowexporter/connections.(*connectionStore).AddConnToMap
    8.50MB  6.86% 59.91%    33.52MB 27.02%  antrea.io/antrea/pkg/agent/flowexporter/exporter.addConnsAndGetRecords
       8MB  6.45% 66.36%    11.50MB  9.27%  github.com/vmware/go-ipfix/pkg/exporter.(*ExportingProcess).createAndSendMsg
       7MB  5.65% 72.00%    15.29MB 12.33%  antrea.io/antrea/pkg/agent/flowexporter/exporter.addDenyConns
    5.50MB  4.43% 76.44%    35.50MB 28.62%  antrea.io/antrea/pkg/agent/flowexporter/exporter.(*flowExporter).addRecordToSet
    4.50MB  3.63% 80.06%     4.50MB  3.63%  github.com/vmware/go-ipfix/pkg/entities.EncodeToIEDataType
       4MB  3.22% 83.29%        4MB  3.22%  github.com/vmware/go-ipfix/pkg/entities.NewDataRecord (inline)
```

```
go tool pprof profile.out
File: exporter.test
Type: cpu
Time: Jun 22, 2021 at 2:01am (UTC)
Duration: 1.24s, Total samples = 1.11s (89.81%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 770ms, 69.37% of 1110ms total
Showing top 10 nodes out of 149
      flat  flat%   sum%        cum   cum%
     510ms 45.95% 45.95%      510ms 45.95%  syscall.Syscall
      60ms  5.41% 51.35%      100ms  9.01%  runtime.scanobject
      40ms  3.60% 54.95%       70ms  6.31%  runtime.mallocgc
      40ms  3.60% 58.56%       40ms  3.60%  syscall.RawSyscall
      20ms  1.80% 60.36%      160ms 14.41%  antrea.io/antrea/pkg/agent/flowexporter/exporter.(*flowExporter).addDenyConnToSet
      20ms  1.80% 62.16%       20ms  1.80%  cmpbody
      20ms  1.80% 63.96%       20ms  1.80%  runtime.heapBits.bits (inline)
      20ms  1.80% 65.77%       40ms  3.60%  runtime.makeslice
      20ms  1.80% 67.57%       20ms  1.80%  runtime.memmove
      20ms  1.80% 69.37%       20ms  1.80%  runtime.step
```
Benchmark result with different number of conntrack connections + deny connections input
| # conns | ns/op | B/op  | allocs/op |
| ------------------------- | ------- | --------- | ----|
| 20,000 + 20,000 | 1,123,878,204 | 135,576,880 |3,458,292|
| 50,000 + 50,000| 2,856,738,645 | 329,267,544 | 8,690,977 |
| 100000 + 100000|5,469,929,206| 656,425,016 | 17,396,982|

zyiou commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,245 @@
+	for i := 0; i < numberOfConn; i++ {
+		randomNum := getRandomNum(b, 255)
+		src := net.ParseIP(fmt.Sprintf("192.168.0.%d", randomNum))
+		dst := net.ParseIP(fmt.Sprintf("192.169.0.%d", randomNum))
```
Actually they are a little different. Made some changes to differentiate them clearer.

zyiou commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,245 @@
+
+var count = 0
+
+func BenchmarkExport(b *testing.B) {
```
Makes sense to me. Updated to a more standard way for benchmarking `Export()` only. Thanks!

zyiou commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,245 @@
+	go func() {
+		defer conn.Close()
+		for {
+			buff := make([]byte, 272)
```
You are right. Here we should use a large number to cover expansion of record length in the future. Changed to a large constant.

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,224 @@
+	}
+
+	for n := 0; n < b.N; n++ {
+		exp.Export()
```
We are considering one set of connections for one export cycle. This is not exercising any code in AddOrUpdateConn method in the conntrack connection store. Is there a plan to have a test for that too?

zyiou replied:
`AddOrUpdateConn` is part of `Poll()` function which will involve mocking dumping flows output from ovs. Do you mean we can test `AddOrUpdateConn` individually or cover testing for `Poll()` too?

srikartati replied:
Covering the test for `Poll()` sounds good as it would include `AddOrUpdateConn`. However, we may need multiple polls as we want to exercise update code or we should initialize the connection store with some connections before the poll. 

I think we are expecting `Run()` method in exporter.go to be exercised in a different e2e test by running it multiple times. Is that correct?

zyiou replied:
Covering test for `Poll()` --- Sure will add that test.
Benchmarking `Run()` in e2e tests --- Yes we should implement it in the future if you are referring to testing with actual polling, updating and exporting.

antoninbas replied:
If possible, let's test things in isolation (when it makes sense) in this type of benchmark

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,224 @@
+			Number of deny connections: 100000
+			Number of idle deny connections: 10000
+			Total connections received: 182861
+	BenchmarkExport-2   	       1	5469929206 ns/op	656425016 B/op	17396982 allocs/op
```
Do you know why we are running the benchmark only one even after adding b.N in the test?
I see the same thing in flow aggregator tests. Not sure what is the reason. 
Should we move to regular test type and track the mem or allocation stats ourselves?
@antoninbas

zyiou replied:
I found if reducing the number of connections, it will run multiple times. It may be dependent on capacity of the machine the test is running on. Here is part of the result:
```
    exporter_perf_test.go:61: 
        Summary:
        Number of conntrack connections: 10000
        Number of dying conntrack connections: 1000
        Number of deny connections: 10000
        Number of idle deny connections: 1000
        Total connections received: 168554
    exporter_perf_test.go:61: 
        Summary:
        Number of conntrack connections: 10000
        Number of dying conntrack connections: 1000
        Number of deny connections: 10000
        Number of idle deny connections: 1000
        Total connections received: 199035
BenchmarkExport-2   	     206	   7352172 ns/op	  492372 B/op	   17727 allocs/op
PASS
ok  	antrea.io/antrea/pkg/agent/flowexporter/exporter	7.627s
```

srikartati replied:
Interesting. Did not know that it depends on the runtime and resource consumption of the test.

antoninbas replied:
It will stop running if one iteration takes too long.

I wonder if we should have different benchmarks for the different types of connections. Would it make sense?

zyiou commented:
After removing redundant calls to `env.GetNodeName()` (latest commit), execution time get reduces by 19.5%.
```
BenchmarkExport
    exporter_perf_test.go:80: 
        Summary:
        Number of conntrack connections: 100000
        Number of dying conntrack connections: 10000
        Number of deny connections: 100000
        Number of idle deny connections: 10000
        Total connections received: 186463
BenchmarkExport-2   	       1	4890531828 ns/op	604259720 B/op	16604105 allocs/op
PASS
ok  	antrea.io/antrea/pkg/agent/flowexporter/exporter	5.258s
```
After change 
```
BenchmarkExport
    exporter_perf_test.go:80: 
        Summary:
        Number of conntrack connections: 100000
        Number of dying conntrack connections: 10000
        Number of deny connections: 100000
        Number of idle deny connections: 10000
        Total connections received: 180096
BenchmarkExport-2   	       1	5096203117 ns/op	603647600 B/op	16591190 allocs/op
PASS
ok  	antrea.io/antrea/pkg/agent/flowexporter/exporter	5.448s
```

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_connections_perf_test.go:
```diff
@@ -0,0 +1,180 @@
+	ok  	antrea.io/antrea/pkg/agent/flowexporter/connections	1.581s
+*/
+
+func BenchmarkPoll(b *testing.B) {
+	disableLogToStderr()
+	setupConntrackConnStore(b)
+	for n := 0; n < b.N; n++ {
+		mockConnDumper.EXPECT().DumpFlows(uint16(openflow.CtZone)).Return(conns, testNumOfConns, nil)
+		connStore.Poll()
+		conns = generateUpdatedConns(conns)
+	}
+	b.Logf("\nSummary:\nNumber of initial connections: %d\nNumber of new connections/poll: %d\nNumber of deleted connections/poll: %d\n", testNumOfConns, testNumOfNewConns, testNumOfDeletedConns)
+}
```
can you try this:
```go
disableLogToStderr()
setupConntrackConnStore(b)
b.ResetTimer()
for n := 0; n < b.N; n++ {
		mockConnDumper.EXPECT().DumpFlows(uint16(openflow.CtZone)).Return(conns, testNumOfConns, nil)
		connStore.Poll()
                 b.StopTimer()
                 conns = generateUpdatedConns(conns)
                 b.StartTimer()
}
```

I feel like the performance of `Poll()` is not that great according to the results, and I wonder if connection generation accounts for a lot of measurement overhead.

Also because of how you designed the benchmark, it looks like we need to run `Poll` multiple times to get good results. So I would recommend reducing the number of connections if needed, to ensure that we have multiple iterations. I'm assuming that resource usage (CPU / memory) increases linearly with the number of connections?

zyiou replied:
Yes, it makes sense to not counting generation of new connections into runtime and run the tests multiple iterations. Tested with different number of connections, CPU and memory are increasing linearly with number of conns:
| # new conns   | # iterations |  ns/op | B/op   | allocs/op |
| ------------------------- | ------- | --------- | --- | --- |
|4000  |    256	|   4738835 ns/op	|  483896 B/op	|   27244 allocs/op| 
|6000  |    206	|   5989825 ns/op	|  608971 B/op	|   36037 allocs/op| 
|8000  |    157	|   7205885 ns/op	|  741305 B/op	|   45061 allocs/op| 
|10000 |   123	|   8710284 ns/op	|  882714 B/op	 |  54307 allocs/op| 
|12000  |   97	|  11834417 ns/op	|  1037488 B/op	|   63868 allocs/op|

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_connections_perf_test.go:
```diff
@@ -0,0 +1,180 @@
+	testNumOfDeletedConns = 10000
+)
+
+var (
+	connStore      *ConntrackConnectionStore
+	conns          []*flowexporter.Connection
+	mockConnDumper *connectionstest.MockConnTrackDumper
+)
```
If possible, avoid global variables here. Wrap these in a struct and populate / return the struct during test setup.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_connections_perf_test.go:
```diff
@@ -0,0 +1,180 @@
+		OriginalBytes:             100,
+		ReversePackets:            5,
+		ReverseBytes:              50,
+		SourcePodNamespace:        "ns1",
+		SourcePodName:             "pod1",
+		DestinationPodNamespace:   "ns2",
+		DestinationPodName:        "pod2",
+		DestinationServiceAddress: net.ParseIP("10.0.0.1"),
+		DestinationServicePort:    30000,
+		TCPState:                  "SYN_SENT",
```
this is a bit misleading IMO. I thought this information was populated by `Poll` itself, it should not be populated in the return value of `DumpFlows`.

zyiou replied:
Yes you are right. Updated.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_connections_perf_test.go:
```diff
@@ -0,0 +1,180 @@
+	ctrl := gomock.NewController(b)
+	defer ctrl.Finish()
+	mockIfaceStore := interfacestoretest.NewMockInterfaceStore(ctrl)
+	mockIfaceStore.EXPECT().GetInterfaceByIP(gomock.Any()).Return(nil, false).AnyTimes()
```
wouldn't it be more accurate to return Pod interface information? I think this can be postponed though, especially if you believe this part of `Poll` (`fillPodInfo`) is not CPU intensive

zyiou replied:
Replaced with Pod interface instead of nil. It turns out that calling to `mockIfaceStore` has some CPU consumption. Maybe we should cache some ip-interface in future work.

antoninbas commented on pkg/agent/flowexporter/connections/conntrack_connections_perf_test.go:
```diff
@@ -0,0 +1,175 @@
+	}
+	randomNum := getRandomNum(int64(length - testNumOfDeletedConns))
+	for i := randomNum; i < testNumOfDeletedConns+randomNum; i++ {
+		updatedConns[i].DoneExport = true
```
what if `updatedConns[i]` is one of the new connections for this iteration? Does it still make sense?

zyiou replied:
In my opinion it is still valid because the connection can be stored and will be deleted in the next round. What do you think?

antoninbas replied:
I think it's ok since we are not testing export here. It would have been good to add a comment to this effect.

antoninbas commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,258 @@
+	exp, err := setupExporter(false)
+	if err != nil {
+		b.Fatalf("error when setting up exporter: %v", err)
+	}
```
The golang testing documentation recommends resetting the timer after doing an "expensive" setup: https://golang.org/pkg/testing/#hdr-Benchmarks

Can you call `b.ResetTimer()` in all you benchmarks, just before the loop?

zyiou replied:
Sure. Thanks!

antoninbas reviewed (commented):
what are the final results, in terms of performance improvements?

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,258 @@
+	testNumOfConns           = 20000
+	testNumOfDenyConns       = 20000
+	testNumOfDyingConns      = 2000
+	testNumOfInactiveRecords = 2000
```
keep the name consistent with the timeout.. `testNumOfIdleRecords`?

srikartati commented on pkg/agent/flowexporter/exporter/exporter_perf_test.go:
```diff
@@ -0,0 +1,258 @@
+	BenchmarkExportConntrackConns-2   	      75	  13750074 ns/op	  965550 B/op	   22268 allocs/op
+	PASS
+	ok  	antrea.io/antrea/pkg/agent/flowexporter/exporter	5.494s
+*/
```
Add the output that you have for the varying number of connections. It will be good to show that.

zyiou replied:
Done. Thanks!

srikartati reviewed (commented):
A couple of nits otherwise LGTM.
Thanks for working on this.

zyiou commented:
> what are the final results, in terms of performance improvements?

Improvement below, also put details in PR description
| Test   |  ns/op | B/op   | allocs/op |
| ------------------------- | --------- | --- | --- |
| BenchmarkExportConntrackConns  |  reduce by 47.4%	|  reduce by 87.6% | reduce by 77.3% | 
| BenchmarkExportDenyConns  |   reduce by 42.9%	|  reduce by 89.6%	|   reduce by 80.0% | 
| BenchmarkPoll  |  less than 1% diff	|  less than 1% diff	| less than 1% diff  |

antoninbas reviewed (dismissed):
LGTM

zyiou commented:
/test-all

srikartati reviewed (approved):
LGTM

antoninbas commented:
/test-e2e
